orbmain/src/main/java/com/sun/corba/ee/impl/protocol/MessageMediatorImpl.java

```diff
@@ -730,85 +726,203 @@ private void connectionInfo( Connection conn ) { }
 
     @Transport
     private void resumeOptimizedReadProcessing(Message message) {
-        messageInfo( message, message.getCorbaRequestId() ) ;
+        messageInfo(message, message.getCorbaRequestId());
         connectionInfo(connection);
 
-        if (message.moreFragmentsToFollow()) {
-            generalMessage("getting next fragment");
-
-            MessageMediator messageMediator = null;
-            RequestId requestId = message.getCorbaRequestId();
-            Queue<MessageMediator> queue =
-                connection.getFragmentList(requestId);
-
-            // REVISIT - In the future, the synchronized(queue),
-            // wait()/notify() construct should be replaced
-            // with something like a LinkedBlockingQueue
-            // from java.util.concurrent using its offer()
-            // and poll() methods.  But, at the time of the
-            // writing of this code, a LinkedBlockingQueue
-            // implementation is not performing as well as
-            // the synchronized(queue), wait(), notify()
-            // implementation.
-            synchronized (queue) {
-                while (messageMediator == null) {
-                    if (queue.size() > 0) {
-                        messageMediator = queue.poll();
-                    } else {
-                        try {
-                            queue.wait();
-                        } catch (InterruptedException ex) {
-                            wrapper.resumeOptimizedReadThreadInterrupted(ex);
-                        }
-                    }
-                }
-            }
-
-            // Add CorbaMessageMediator to ThreadPool's WorkQueue to process the
-            // next fragment.
-            // Although we could call messageMeditor.doWork() rather than putting
-            // the messageMediator on the WorkQueue, we do not because calling
-            // doWork() would increase the depth of the call stack. Since this
-            // thread is done processing the Work it was given, it is very likely
-            // it will be the thread that executes the Work (messageMediator)we
-            // put the on the WorkQueue here.
-            addMessageMediatorToWorkQueue(messageMediator);
+        if (message.moreFragmentsToFollow() && !isNewFragmentProcessingSet) {
+            synchronizedProcess(message);
+        } else if (message.moreFragmentsToFollow() && isNewFragmentProcessingSet) {
+            lockProcess(message);
         } else {
```

Comment:
Could be cleaned up slightly

```suggestion
        if (message.moreFragmentsToFollow()) {
            if (!isNewFragmentProcessingSet) {
                synchronizedProcess(message);
            } else {
                lockProcess(message);
            }
        } else {
```

Reply:
already done